src/olympia/devhub/tests/test_tasks.py

```diff
@@ -616,6 +611,63 @@ def test_amo_validator_addons_linter_error(self):
         assert not result['warnings']
 
 
+class TestWebextensionUpgrade(TestCase):
+    fixtures = ['base/addon_3615']
+
+    def setUp(self):
+        self.addon = Addon.objects.get(pk=3615)
+
+        # valid_webextension.xpi has version 1.0 so mock the original version
+        self.addon.update(guid='beastify@mozilla.org')
+        self.addon.current_version.update(version='0.9')
+        self.update_files(
+            version=self.addon.current_version,
+            filename='delicious_bookmarks-2.1.106-fx.xpi')
+
+    def update_files(self, **kw):
+        for version in self.addon.versions.all():
+            for file in version.files.all():
+                file.update(**kw)
+
+    def test_webextension_upgrade_is_annotated(self):
+        assert all(f.is_webextension is False
+                   for f in self.addon.current_version.all_files)
+
+        file_ = get_addon_file('valid_webextension.xpi')
+        upload = FileUpload.objects.create(path=file_, addon=self.addon)
+
+        tasks.validate(upload)
+
+        upload.refresh_from_db()
+        assert upload.processed_validation['is_upgrade_to_webextension']
+
+        expected = ['validation', 'messages', 'webext_upgrade']
+        assert upload.processed_validation['messages'][0]['id'] == expected
+
+    def test_webextension_webext_to_webext(self):
+        previous_file = self.addon.current_version.all_files[-1]
+        previous_file.is_webextension = True
+        previous_file.save()
+
+        file_ = get_addon_file('valid_webextension.xpi')
+        upload = FileUpload.objects.create(path=file_, addon=self.addon)
+
+        tasks.validate(upload)
+        upload.refresh_from_db()
+
+        assert 'is_upgrade_to_webextension' not in upload.processed_validation
+
+    def test_webextension_no_webext_no_warning(self):
```

Comment:
so, we warn developers that once they go webext they can't go back... but if they do go back to xul we don't warn them? Or is that check elsewhere?